File: common/src/main/java/org/valkyrienskies/clockwork/integration/cc/PhysBearingPeripheral.java
package org.valkyrienskies.clockwork.integration.cc;

import com.simibubi.create.foundation.tileEntity.behaviour.scrollvalue.ScrollOptionBehaviour;
import dan200.computercraft.api.lua.IArguments;
import dan200.computercraft.api.lua.LuaException;
import dan200.computercraft.api.lua.LuaFunction;
import dan200.computercraft.api.peripheral.IPeripheral;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.valkyrienskies.clockwork.content.contraptions.phys.bearing.PhysBearingBlockEntity;
import org.valkyrienskies.clockwork.platform.api.ContraptionController;

import java.util.Optional;

public class PhysBearingPeripheral implements IPeripheral {
    private final PhysBearingBlockEntity phys;

    public PhysBearingPeripheral(PhysBearingBlockEntity be) {
        this.phys = be;
    }

    @NotNull
    @Override
    public String getType() {
        return "phys_bearing";
    }

    @Override
    public boolean equals(@Nullable IPeripheral iPeripheral) {
        return iPeripheral instanceof PhysBearingPeripheral;
    }

    @LuaFunction
    public final void setAngle(double angle) {
        this.phys.setAngle((float) angle);
    }

    @LuaFunction
    public final double getAngularSpeed() {
        return this.phys.getAngularSpeed();
    }

    @LuaFunction
    public final void assemble() {
        this.phys.assemble();
    }

    @LuaFunction
    public final void disassemble() {
        this.phys.disassemble();
    }

    @LuaFunction
    public final boolean isRunning() {
        return this.phys.isRunning();
    }

    @LuaFunction
    public final void release() {
        this.phys.destroy();
    }

    @LuaFunction
    public final double getAngle() {
        return this.phys.getAngle();
    }

    @LuaFunction
    public final void lock(IArguments args) throws LuaException {
        ScrollOptionBehaviour<ContraptionController.LockedMode> mode = this.phys.getMovementMode();
        boolean locked = args.optBoolean(0, !mode.get().equals(ContraptionController.LockedMode.LOCKED));

        mode.setValue(locked ? 1 : 0);

        this.phys.setMovementMode(mode);
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/integration/cc/PhysBearingPeripheral.kt
@file:Suppress("unused")

package org.valkyrienskies.clockwork.integration.cc

import dan200.computercraft.api.lua.LuaFunction
import dan200.computercraft.api.peripheral.IPeripheral
import net.minecraft.core.BlockPos
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import org.valkyrienskies.clockwork.ClockworkBlocks
import org.valkyrienskies.clockwork.content.contraptions.phys.bearing.PhysBearingBlockEntity
import org.valkyrienskies.clockwork.platform.api.ContraptionController
import org.valkyrienskies.mod.common.getShipManagingPos

class PhysBearingPeripheral(private val be: PhysBearingBlockEntity): IPeripheral {
    @LuaFunction fun assemble() = be.assemble()
    @LuaFunction fun disassemble() = be.disassemble()

    @LuaFunction fun setFollowAngleMode() {
        be.movementMode!!.setValue(ContraptionController.LockedMode.FOLLOW_ANGLE.ordinal)
    }
    @LuaFunction fun setUnlockedMode() {
        be.movementMode!!.setValue(ContraptionController.LockedMode.UNLOCKED.ordinal)
    }
    @LuaFunction fun setAngle(angle: Double) {
        be.setAngle(angle.toFloat())
    }

    @LuaFunction fun isBeingDisassembled() = be.aligning
    @LuaFunction fun isActive() = be.isRunning
    @LuaFunction fun isInFollowAngleMode() = be.movementMode!!.get() == ContraptionController.LockedMode.FOLLOW_ANGLE

    @LuaFunction fun getConnectedToShip() = be.level.getShipManagingPos(be.partnerPos ?: BlockPos(0, 0, 0))?.id ?: -1
    @LuaFunction fun getTargetAngle() = be.targetAngle
    @LuaFunction fun getActualAngle() = be.getActualAngle()

    @LuaFunction fun getRPM() = be.speed
    @LuaFunction fun getFacingDirection() = be.blockState.getValue(BlockStateProperties.FACING).getName()

    override fun equals(p0: IPeripheral?): Boolean = be.level?.getBlockState(be.blockPos)?.`is`(ClockworkBlocks.PHYS_BEARING.get()) == true
    override fun getType(): String = "cw_phys_bearing"
}